src/ripple/shamap/impl/SHAMap.cpp

```diff
@@ -78,6 +83,72 @@ SHAMap::snapShot (bool isMutable) const
     return ret;
 }
 
+std::shared_ptr<SHAMap>
+SHAMap::make_v2() const
+{
+    assert(!is_v2());
+    auto ret = std::make_shared<SHAMap>(type_, f_, version{2});
+    ret->seq_ = seq_ + 1;
+    SharedPtrNodeStack stack;
+    for (auto leaf = peekFirstItem(stack); leaf != nullptr;
+         leaf = peekNextItem(leaf->peekItem()->key(), stack))
+    {
+        auto node_type = leaf->getType();
+        ret->addGiveItem(leaf->peekItem(),
+                         node_type != SHAMapTreeNode::tnACCOUNT_STATE,
+                         node_type == SHAMapTreeNode::tnTRANSACTION_MD);
+    }
+    NodeObjectType t;
+    switch (type_)
+    {
+    case SHAMapType::TRANSACTION:
+        t = hotTRANSACTION_NODE;
+        break;
+    case SHAMapType::STATE:
+        t = hotACCOUNT_NODE;
+        break;
+    default:
+        t = hotUNKNOWN;
+        break;
+    }
+    ret->unshare();
+    ret->flushDirty(t, ret->seq_);
+    return ret;
+}
+
+std::shared_ptr<SHAMap>
+SHAMap::make_v1() const
```

Comment:
Under `SHAMap::fetchNodeFromDB` we read nodes from a database.  We can get into a situation where we have switched to v2, but we read a v1 shamap out of a database.  We need to be able to switch back to v1 mode to read that data.  There's also a similar use case under `SHAMap::checkFilter`.